Provide clear guidance on what semconv actually enforces (#3225)
open-telemetry/opentelemetry-specification: specification/versioning-and-stability.md

relies on the concept of
[Telemetry Schemas](schemas/README.md).

Semantic Conventions are not enforced on all fields available for telemetry.
Instead, semantic conventions apply to "API" type interfaces on telemetry as
opposed to actual values. Explicitly, Semantic conventions apply to the
following API calls when constructing instrumentation:

- [Resource](resource/sdk.md)
  - attribute keys provided to Create or resource detectors

> - [Trace](trace/api.md)

Reviewer: Some semantic conventions for traces also contain conventions for [span kind](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#span-kind) and [span status](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#status) values. I wonder what this proposal entails for those "conventions" not covered by this list:

* Are they supposed to stay as part of semantic conventions, but are they not covered by stability guarantees? I guess in that case, they would just be recommendations?
* Should those be removed from the current semantic conventions, if this proposal is accepted?
* Should they be added to this list?

Span kind and status are interesting cases in this regard, as the specification explicitly refers to semantic conventions for both.

For [span kind](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind):
> Note: there are complex scenarios where a CLIENT span may have a child that is also logically a CLIENT span, or a PRODUCER span might have a local child that is a CLIENT span, depending on how the various libraries that are providing the functionality are built and instrumented. These scenarios, when they occur, should be detailed in the semantic conventions appropriate to the relevant libraries.

For [span status](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-status):
> When the status is set to Error by Instrumentation Libraries, the Description SHOULD be documented and predictable. The status code should only be set to Error according to the rules defined within the semantic conventions. For operations not covered by the semantic conventions, Instrumentation Libraries SHOULD publish their own conventions, including possible values of Description and what they mean.

Author: I'll answer the easy one first:

Span status is un-enforcable.  We can provide guidance on what to consider an error, but we cannot enforce anything about it in practice.  E.g. the same "span API" may have error spans or successful spans, as the status is a runtime-computed value.  This is similar, e.g. to metric point values which are calculated at runtime.  They are outside the scope of an enforcable "stable" API.

For Span Kind, that's a good call-out.  I think that was oversight on my part. I believe span kind / span-name should be enforced similarly.  We *can* ensure that if a "span-name foo" shows up, it has a particular kind and set of attribute keys.

Author: Updated, PTAL

Reviewer: It's still not fully clear to me what this entails for current parts of semantic conventions that do not fit into this list, like assumptions about [span status](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#status):

> [Span Status](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-status) MUST be left unset if HTTP status code was in the 1xx, 2xx or 3xx ranges, unless there was another error (e.g., network error receiving the response body; or 3xx codes with max redirects exceeded), in which case status MUST be set to Error.

Or strong statements about attribute values like this one:

> `net.sock.family` has the following list of well-known values. If one of them applies, then the respective value MUST be used, otherwise a custom value MAY be used.

Both examples use MUST language but fall under items in the list below as "[...] not enforced via semantic convention and allowed (or expected) to change".

Should those things be removed from semantic conventions? Or should they still be part as recommendations that aren't enforced and aren't subject of stability guarantees? Or can semantic conventions still _require_ things although they cannot be _enforced_?

Author: I added some clarifying language. PTAL

Reviewer: That makes sense, thanks.

If I understand correctly, as a consequence, (stable) semantic conventions will then have a "strict" part that can be covered by schemas and schema conversions, and another part containing recommendations and suggestions (with no "MUST" language).

Author: That's my proposal, yes.

Author: Marking this resolved, please reopen if unaddressed.

  - Attribute keys provided to [get a tracer](trace/api.md#get-a-tracer)
  - The following data provided to [span](trace/api.md#span) by available APIs:
    - The span name
    - The attribute keys provided to the span.

>   - The following data provided when [adding an event](trace/apid.md#add-events)

Reviewer: ```suggestion
  - The following data provided when [adding an event](trace/api.md#add-events)
```

    - The event name
    - The attribute keys provided for the event.
- [Metrics](metrics/api.md)
  - Attribute keys provided to [get a meter](metrics/api.md#get-a-meter)
  - The following arguments when constructing [an instrument](metrics/api.md#instrument):
    - The name of the instrument.
    - The kind of instrument.
      - For `Counter` and `UpDownCounter` instruments, it is

>         acceptable to use asynchronous or synchronous instruments.

Reviewer 2: ```suggestion
        acceptable to change between asynchronous and synchronous instruments.
```

    - the unit of the instrument.
  - The attribute keys provided when recording a measurement, for
    both synchronous and asynchronous instruments.

> - Logs (via [bridge](logs/bridge-api.md) or [event](logs/event-api.md) are not currently enforced.

Reviewer 2: ```suggestion
- Logs (via [bridge](logs/bridge-api.md) or [event](logs/event-api.md)) are not currently enforced.
```

relies on the concept of
[Telemetry Schemas](schemas/README.md).

> Semantic Conventions are not enforced on all fields available for telemetry.

Reviewer 3: I find this odd - semantic **conventions** are recommendations, they are not enforced anywhere at all.

Reviewer 2: maybe something like?
```suggestion
Semantic convention stability guarantees only apply to specific parts of the telemetry data.
```

Author: Stability guarantees that come along with Semantic conventions and declaring your support of them with Schema URL.  I'll update the language to represent this.

Author: I updated the description here, PTAL and resolve if the new verbage works.

> Semantic Conventions are not enforced on all fields available for telemetry.

Reviewer 4: I don't understand what this section is trying to achieve. I think it is rather clear from the semantic conventions themselves what they apply to. Are there any things we want to explicitly exclude? Why?

Author: What they enforce actually isn't clear, as we've learned in the semconv WG.  If you ask different people across OTEL you'll see different ideas on what constitutes a breaking or non-breaking change.

This is an attempt to be 100% specific on what is enforced to avoid ambiguity and misunderstanding going forward.

I think if you look at the discourse in this PR, you'll see the nuances of things we wish to enforce in semconv and things we need to consider out of scope or guidance only.

Reviewer 4: To me it isn't clear what "enforce" means, so this PR is very confusing to me. If this is about defining what constitutes a breaking change, then I would use different terms like "for semantic conventions, the following changes constitute a breaking change ...". "Breaking changes are only allowed if ... and require entries in the schema and ..."

Author: This is about defining what constitutes a breaking change in semconv, and which fields can have any kind of enforced stability vs. ones that can/will change.

I'm not sure I understand your suggestion here or how to word it in the scope of the document.

Author: I took a crack at rewording this based on your + Tigran's comments, PTAL.

Author: Please reopen if this is unaddressed

Instead, semantic conventions apply to "API" type interfaces on telemetry as
opposed to actual values. Explicitly, Semantic conventions apply to the
following API calls when constructing instrumentation:

- [Resource](resource/sdk.md)
  - attribute keys provided to Create or resource detectors
- [Trace](trace/api.md)
  - Attribute keys provided to [get a tracer](trace/api.md#get-a-tracer)
  - The following data provided to [span](trace/api.md#span) by available APIs:

>     - The span name

Reviewer 3: Guidelines for span name are defined in the spec itself, not in the semantic conventions.

Reviewer 2: some semantic conventions have their own more specific span name guidelines, e.g. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#name

Author: As noted elsewhere, there are additional span name considerations added by semantic conventions.

Author: Updated the description. PTAL.

Author: Reopen if unaddressed, please.

>     - The span kind

Reviewer 3: Isn't span kind enforced via enum / API? It's not a free-form text.

Reviewer 2: I think the idea for span kind being part of stability guarantee is to say we can't change the HTTP semantic conventions from

> For an HTTP client span, `SpanKind` MUST be `Client`.

to 

> For an HTTP client span, `SpanKind` MUST be `Producer`.

(or something like that)

Author: This is suggesting that once I declare a particular span needs to be emitted as a 'client' I can't change it to 'server' later.

>     - The attribute keys provided to the span.

Reviewer 5: also

+ when attributes are provided (at start time or after it)
+ attribute requirement levels for required and conditionally-required attributes

Author: How are we enforcing when attributes are provided?  Is that exposed in the YAML today?

I realize this is due to interactions w/ sampling, but I want to make sure what we enforce/say is enforcable.

Regarding attribute requirement levels, that's a separate specification that discusses this.

Reviewer 5: we do have `sampling_relevant` flag https://github.com/open-telemetry/opentelemetry-specification/blob/dd47c009da1de335c540decc5d5930e46ea61916/semantic_conventions/trace/http.yaml#L14

and tooling that populates the note like this:

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md?plain=1#L163-L167

So it seems we're ready to say it's enforceable.

Author: Good point, I'll add this.

  - The following data provided when [adding an event](trace/api.md#add-events)

>     - The event name

Reviewer 3: afaict the event name is completely free-form, no guidance around it.

Reviewer 2: some semantic conventions have their own more specific event name guidelines, e.g. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/rpc.md#events

Author: We have some semantic conventions recommending event names.

Author: Closing this comment, please reopen if you feel this needs to be addressed in this PR.

    - The attribute keys provided for the event.
- [Metrics](metrics/api.md)
  - Attribute keys provided to [get a meter](metrics/api.md#get-a-meter)
  - The following arguments when constructing [an instrument](metrics/api.md#instrument):
    - The name of the instrument.
    - The kind of instrument.
      - For `Counter` and `UpDownCounter` instruments, it is
        acceptable to change between asynchronous and synchronous instruments.
    - the unit of the instrument.
  - The attribute keys provided when recording a measurement, for
    both synchronous and asynchronous instruments.

> - Logs (via [bridge](logs/bridge-api.md) or [event](logs/event-api.md)) are not currently enforced.

Reviewer 6: I suggest that we remove anything that is not enforced:
1. So we don't have to keep updating this document when new things show up
2. So we don't run into weird situation of something not listed here and become something that is neither "enforced" nor "not enforced"

Reviewer 6: ```suggestion
- For anything else that hasn't been listed above, they are not currently enforced.
```

Author: This is my reference that we need to add logs enforcement.  I can take a crack in htis PR, but it's important to know it's not done.

Reviewer 6: Got it, maybe "anything that is not listed above, is not currently enforced (e.g. logs, ...)"

Reviewer 7: The event piece is wrong. There is strong intent to have events defined with semantic conventions. This includes the [attributes which identify that a log record is an event](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/semantic_conventions/events.md), and (eventually) the schema of events (#2933 went stale but attempted to define semantic conventions for specific RUM events).

Reviewer 8: As jack mentioned, the intention is that for each defined `event` (domain / name combination) there will be a define set of semantic conventions for the event payload (which will be included in the `event.data` as being defined in #2926 -- update to better highlight are coming)

Author: @jack-berg Can you clarify what's wrong?

1. We can provide what we consider stable in events later, given the language.
2. Is there something actively wrong with what's listed on Span Events today?

Reviewer 7: > We can provide what we consider stable in events later, given the language.

I'm trying to understand what we accomplish by omitting events in this version of the document?

Author: > I'm trying to understand what we accomplish by omitting events in this version of the document?

The ability to argue nuances in a separate PR and make slow progress.

I'll add events if you feel it belongs in a big-bang change.

Reviewer 7: Yes, it is strange that the exception semantic conventions include language about the API definition

> To encapsulate proper handling of exceptions API authors MAY provide a constructor, RecordException method/extension, or similar helper mechanism on the LogRecord class/structure or wherever it makes the most sense depending on the language runtime.

Seems like a mistake.

Looks good, other than the omission of event domain I mentioned [here](https://github.com/open-telemetry/opentelemetry-specification/pull/3225#discussion_r1125036282).

Reviewer 7: I believe this is resolved. 👍

Things not listed in the above are not enforced via semantic convention and allowed (or expected) to change. A few examples:

> - The values of attributes

Reviewer 5: what about types? converting string to array or int would be breaking

Author: So the *attribute keys* are in scope, and the key defines the type of attribute.

Yes, We need to consider type changes breaking.

Reviewer 9: In our terms I think the attribute key is just the name of the attribute, which does not include the type. I think the type should be mentioned explicitly and indeed considered breaking if changed.

Reviewer 9: If we define an enum that does not allow for custom values, then I think this is considered its own (stable) type as telemetry consumers will only expect the values listed there to appear in the data they receive. The consequence of that would be that adding new values or opening the enum up would be breaking for them as well. If we don't want that, we'd need to call this out explicitly.

Author: I'm ok adding language specific to enums if we think we have enough in place to define and enforce stability around them today.

> - The links attached to a span

Reviewer 5: semconvs need to be able to enforce links and relationships between spans.
E.g. in case of messaging, it's necessary to describe how per-message attributes are populated when a batch of messages is being sent.

I'm fine with leaving it as it is for now and returning back to it once we're ready to stabilize messaging semconv.

Author: This is why I want this document. I don't think we're prepared to enforce links at this point.

We need a way to define in semconv YAML:
- link attributes (easy fix)
- link semconvs that apply to a Span semconv

I'd rather expliciltly leave this out of scope until we have better machinery to deal with it.

relies on the concept of
[Telemetry Schemas](schemas/README.md).

> Semantic Conventions enforce stability in the telemetry produced for tooling.

Reviewer 10: In what sense is the word "enforce" used here? We don't enforce the conventions in the Otel code in any way currently. We provide helpers so that you can generate telemetry according to the semantic conventions but the developers are not forced to use the helpers.

Author: Reworded this section, PTAL.  I've removed the word enforce due to common confusion it caused.

Semantic Convention stability applies to "API" type interfaces on telemetry as

> opposed to actual values. Semantic Conventions are allowed to recommend and

Reviewer 10: >Semantic Convention stability applies to "API" type interfaces on telemetry as
>opposed to actual values.

I am not sure I understand what this means. What is `"API" type interface`? Which "actual values"?

Author: Updated the description, PTAL.

This was about focusing on values a tooling provider would interact with (API) vs. the dynamic values you expect to change.

Author: Please reopen if the new description doesn't work.

suggest the shape of runtime values. However only the following fields will have

> enforced stability and interaction with telemetry schemas:

Reviewer 10: The meaning of "and interaction with telemetry schemas" is not quite clear.

Author: Removed and left for the other sections to outline this.

- [Resource](resource/sdk.md)

>   - attribute keys provided to Create or resource detectors

Reviewer 10: For some attributes we also define _values_ to be used, e.g. for `telemetry.sdk.language` the value can be one from the specified enumeration or a custom value.
The same applies to span/metric/log attributes.

Author: I agree we tried to define values in semconv.  I'm suggesting we shouldn't include those yet *or* that we need to more aggressively build features to ensure stability and evolution of those definitions.

- How do these value specifications interact w/ telemetry schemas?  Can we provide migrations like we do for attribute names?
- Are you able to expand the definition of what values are allowed to include more in a non breaking way?
- How do we enforce values conform to the specification?

I'm suggesting that basically all "value" semconv is limited to SHOULD, and stability wise, tooling should expect new values to show up.

If you want to limit it further, then we'll need to refine *which* values can be locked down, expand telemetry schema to interact with values *and* update the metadata yaml to denote attribute format requirements, ideally, in a way we can programatically enforce via generated tooling.

Author: I'm marking this as closed, please reopen if my comment didn't address your concern.

- [Trace](trace/api.md)
  - Attribute keys provided to [get a tracer](trace/api.md#get-a-tracer)
  - The following data provided to [span](trace/api.md#span) by available APIs:
    - The span name
    - The span kind
    - The attribute keys provided to the span.
  - The following data provided when [adding an event](trace/api.md#add-events)
    - The event name
    - The attribute keys provided for the event.
- [Metrics](metrics/api.md)
  - Attribute keys provided to [get a meter](metrics/api.md#get-a-meter)
  - The following arguments when constructing [an instrument](metrics/api.md#instrument):
    - The name of the instrument.
    - The kind of instrument.
      - For `Counter` and `UpDownCounter` instruments, it is
        acceptable to change between asynchronous and synchronous instruments.
    - the unit of the instrument.
  - The attribute keys provided when recording a measurement, for
    both synchronous and asynchronous instruments.

> Things not listed in the above are not enforced via semantic convention and allowed (or expected) to change. A few examples:

Reviewer 10: ```suggestion
Things not listed in the above are not enforced via semantic convention and are allowed (or expected) to change. A few examples:
```

> - The values of attributes

Reviewer 10: This is generally speaking incorrect. For some attributes we define what possible values can be used, what the values should look like, etc.

Author: We don't really have a way to enforce those values, only their meaning.

Reviewer 4: What does "enforce" mean... I think its good that the PR wording has changed now. Of course changing the value to something with a different meaning should constitute a breaking change IMHO.

Author: Reworded, PTAL

Author: Please reopen in unaddressed.

relies on the concept of
[Telemetry Schemas](schemas/README.md).

Semantic Conventions defines breaking changes as those that would break the

> "API" of tooling written against the telemetry it produces. That is the portions

Reviewer 2: ```suggestion
"API" of tooling written against the telemetry it produces. That is, the portions
```

of telemetry where specialized tooling (alerts, dashboards, e.g.) interact are
expected to remain stable for that tooling. Semantic Conventions defines the

> set of fields to which stability concerns apply are as follows:

Reviewer 2: ```suggestion
set of fields to which stability concerns apply as follows:
```

- [Resource](resource/sdk.md)

>   - attribute keys provided to Create or resource detectors

Reviewer 4: I think we should not describe this in terms of the specific APIs that might be used to set these attributes, as it may easily miss specific APIs. Also I don't think we plan to exclude any APIs to leave a backdoor where no stability concerns apply. Instead, we should describe this in terms of our data model. I think this is already done partially, but e.g. for Resources, I would simply list "'Resources", since they don't have any complex data model and it is rather irrelevant (even transparent) to receivers, how the resource attributes were provided.

Author: I think agree with this.  

However, the wrench here is when the user has access between API->Export to DataModel.  E.g. Metric Views.  To some extent, the semconv should outline what the instrumentation will provide *barring no overrides*.  I attempted to then focus on the API instrumentation would be using, but it did get awkward.

I can migrate this to discuss the datamodel, but does "OTLP produced in lieu of any external user configuration" sound understandable?

Reviewer 4: OTLP sounds good to me as "reference data model", since I don't think we have a explicit prescriptive data model definition otherwise.

But I also don't think it would be too bad to just use commonly understood terms like "event attributes", "span attributes". I don't think this is a point of much ambiguity.

Reviewer: That's a great point, I agree with @Oberon00. 

In my opinion, a user that carefully overrides or enriches telemetry data so that it stays in line with semantic conventions and schema versions should get stability guarantees too.

That's also how I understand the introductory paragraph:
> Semantic Conventions defines breaking changes as those that would break the "API" of tooling written against _the telemetry it produces_.

The produced telemetry is a result of both instrumentation, user overrides, and possibly other transformations. To avoid break tooling, all of those need to create and modify telemetry data according to semantic conventions.

Reviewer 2: > In my opinion, a user that carefully overrides or enriches telemetry data so that it stays in line with semantic conventions and schema versions should get stability guarantees too.

this makes me wonder if we should reduce some `Required` http metric attributes to `Recommended`, in order to allow users to drop them based on their cardinality needs without violating HTTP semconv, e.g.

* `net.peer.name` is required on HTTP client metrics, which could be problematic for some use cases
* `http.status_code` is in a similar boat, see #2947

maybe we need 2 different attribute requirement levels, one for instrumentation (where we may want Required) and one for telemetry consumers (where we may want Recommended)?

Author: Ok, I updated the description to be OTLP based, with caveats on API usage for instrumentation authors. PTAL.

>   - attribute keys provided to Create or resource detectors

Reviewer 2: ```suggestion
  - attribute keys provided to Create resource detectors
```

- [Trace](trace/api.md)
  - Attribute keys provided to [get a tracer](trace/api.md#get-a-tracer)
  - The following data provided to [span](trace/api.md#span) by available APIs:
    - The span name
    - The span kind
    - The attribute keys provided to the span
      - Whether these attributes must be provided at span start time, due to
        sampling concerns.
  - The following data provided when [adding an event](trace/api.md#add-events)
    - The event name

>     - The attribute keys provided for the event

Reviewer 4: What about the values? I think instead of "attribute keys" it should say "attributes". (Of course the concrete values will change but the definition of what the values are (e.g. does the URL include any query) should be covered by stability, not just the keys)

Reviewer 4: > without also falling into the trap of, e.g. allowing enumerations to be "locked".

I actually think all enumerations should be locked by default if they don't allow custom values. But enums are probably a common-enough point to state explicitly: "Adding new members to enum attributes that don't allow custom values constitutes *a/no* breaking change.

Reviewer 5: lack of enums back-compat and future-proofness in languages like Java hurts a lot making people come up with better extensible enum-like implementations.

Trying to enforce anything across multiple languages and technologies that don't even exist yet (e.g. yet another messaging system that will pop up tomorrow)  is problematic.

I think what enforcement means in scope of this document is "what telemetry consumers can expect to change" rather than "what instrumentations should do"

E.g.  we should recommend instrumentations use one of the available values in the enum and even might require it in some cases, but consumers MUST expect custom values in the enums if they want to support future versions of semconv.

Reviewer 4: >  yet another messaging system that will pop up tomorrow

Obviously you would use allow_custom_values there. I think allow_custom_values should probably be used in most cases and allow_custom_values=false should be used only very sparingly when the set of possible enumerators is logically finite and closed.

Author: I firmly believe the opposite.  I'm against having `allow_custom_values` in semantic conventions and think we should remove it.  If we allow it, we need to be VERY careful in how we use it.

Reviewer 11: How can/should we resolve this?

Reviewer 4: I have no strong opinion here, so if you want to resolve this, you can go with what @jsuereth suggests. To implement this you need IMHO:
* a PR to the build tools to change the wording it generates for enums and remove the allow_custom_values option
* a PR to the spec to remove all usages of allow_custom_values
* a clarification dedicated to enums here in this PR like "In particular, the type of an attribute may be changed, e.g. from string to enum, or by adding or removing enum members"

Author: Thanks @Oberon00.  I opened #3359 to track this work.

Regarding point three:  There is no such type as "enum" in OTLP.  "enum" was always used as a convenience for string types.  I think we need a caveat there, but from a specification language I think we're ok here.

- [Metrics](metrics/api.md)
  - Attribute keys provided to [get a meter](metrics/api.md#get-a-meter)
  - The following arguments when constructing [an instrument](metrics/api.md#instrument):
    - The name of the instrument
    - The kind of instrument
      - For `Counter` and `UpDownCounter` instruments, it is
        acceptable to change between asynchronous and synchronous instruments.
    - the unit of the instrument.
  - The attribute keys provided when recording a measurement, for
    both synchronous and asynchronous instruments.
- [Log Records](logs/bridge-api.md#logrecord)
  - The attribute keys provided on the LogRecord
- [Log Events](logs/event-api.md)
  - The following data provided to [emit event](logs/event-api.md#emit-event):

>     - The event name

Reviewer 7: ```suggestion
    - The event name
    - The event domain
```

    - The components of [LogRecord](logs/bridge-api.md#logrecord) also
      listed here.

Things not listed in the above are not enforced via semantic convention and are allowed (or expected) to change. A few examples:

- The values of attributes
  - Specifically for `enums` the list of allowed values is expected to change
    overtime.

>   - Event for `enums` that limit allowed values to semconv, some may need to

Reviewer 2: ```suggestion
  - Even for `enums` that limit allowed values to semconv, some may need to
```

    updated values in the future.
- The links attached to a span
- The recorded measurement type (float or integer) of a metric is not enforced and allowed to change.
- The description of a metric instrument.
- The values being recorded by an instrument.

> The list of telemetry fields with enforced stability MAY be appended.

Reviewer 4: How will stability be enforced? Probably this is meant.

```suggestion
The list of telemetry fields which are covered by stability guarantees MAY be extended.
```

(NB: In my reading, the MAY allows every SIG to define additional fields for which they provide stability guarantees. I'm not sure if that is a good idea, maybe this was meant as non-normative text like this instead:

```suggestion
Note: The list of telemetry fields which are covered by stability guarantees should be expected to be extended in future versions of the spec.
```

)

relies on the concept of
[Telemetry Schemas](schemas/README.md).

Semantic Conventions defines breaking changes as those that would break the
"API" of tooling written against the telemetry it produces. That is, the portions
of telemetry where specialized tooling (alerts, dashboards, e.g.) interact are
expected to remain stable for that tooling. Semantic Conventions defines the
set of fields to which stability concerns apply as follows:

- [Resource](resource/sdk.md)

>   - attribute keys provided to Create resource detectors

Reviewer 4: I don't think this is a contradiction. Stability concerns(!) apply, and the concerns are resolved in the case of renames by providing a schema entry.

>   - attribute keys provided to Create resource detectors

Reviewer 5: agree, it's not a contradiction but rather confusion as this sections defines instrumentation stability guarantees and schema transformation applies to consumers. It's a similar discussion as in https://github.com/open-telemetry/opentelemetry-specification/pull/3289 if we're ready to provide consumer guarantees and what they are. 

Anyway, I don't think we should block this PR on this and can continue in https://github.com/open-telemetry/opentelemetry-specification/issues/3296

    - The attribute keys provided to the span
      - Whether these attributes must be provided at span start time, due to
        sampling concerns.
  - The following data provided when [adding an event](trace/api.md#add-events)
    - The event name
    - The attribute keys provided for the event
- [Metrics](metrics/api.md)
  - Attribute keys provided to [get a meter](metrics/api.md#get-a-meter)
  - The following arguments when constructing [an instrument](metrics/api.md#instrument):
    - The name of the instrument
    - The kind of instrument
      - For `Counter` and `UpDownCounter` instruments, it is
        acceptable to change between asynchronous and synchronous instruments.
    - the unit of the instrument.
  - The attribute keys provided when recording a measurement, for
    both synchronous and asynchronous instruments.
- [Log Records](logs/bridge-api.md#logrecord)
  - The attribute keys provided on the LogRecord
- [Log Events](logs/event-api.md)
  - The following data provided to [emit event](logs/event-api.md#emit-event):
    - The event name
    - The event domain
    - The components of [LogRecord](logs/bridge-api.md#logrecord) also
      listed here.

> Things not listed in the above are not enforced via semantic convention and are allowed (or expected) to change. A few examples:

Reviewer: If I understand correctly, these examples mention parts of telemetry that are not covered by stability guarantees, but that can still be covered by semantic conventions (as recommendations and suggestions)?

If that's the case, the wording here ("are allowed (or expected) to change") contradicts wording below:

> Changes to semantic conventions in this specification are allowed, provided that the changes can be described by schema files.

If that's not the case, it would be helpful to make clear that things in the list below aren't just "not enforced via semantic conventions", but actually are "not part of semantic conventions".

There are "semantic conventions", and there are "things to which stability guarantees apply", and it's not clear to me whether we want both to mean the same, or whether there's just an overlap.

Author: I'm trying to address both these problems.  Let me rephrase to make this clear.  The goal is though:

1. Tooling authors understand what they should expect to change over time and what will remain stable.
2. Semconv provides guidelines on meaning + terminology for instrumentation.  We should be able to validate whether any instrumentation is abiding by semconv via what's specified.

Author: updated description again, PTAL.

described) by schema files. Here is the list of such changes:

- Adding new attributes to the existing semantic conventions for resources,
  spans, span events or log records.

> - Adding new attributes to existing metrics that do not create new timeseries.

Reviewer 6: I find this line hard to understand - does this mean "introducing new attributes that are off-by-default (and can be opted-in) to existing metrics"?

Author: This one trips a few people.  I need to add an example.

Reviewer 2: can you also expand on why this would break alerts (if a new attribute does create a new timeseries)?

e.g. my naive assumption is that an alert would be something like "p(95) where http.route = login" and that backends would aggregate across all the other dimensions

Author: I think the secret here is that for timeseries queries, generally, aggregation is on-demand, not implicit.

E.g. a promql query for `http.duration` (should PromQL support `.`) would return EVERY timeseries.  So something like `histogram_quantile(0.95, sum(rate(http_request_duration_seconds_bucket[5m])) by (le))` would preserve every timeseries by default.

Author: I updated the wording here, PTAL

Author: Reopen if unanswered, please.
